[ncduy0303] iP

ncduy0303 opened the pull request:
### Meet Alice

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Allow Alice to become your personal assistant, liberating your mind from the burdensome task of remembering all the things you have to do. With Alice, managing your tasks is,

- Text-based
- Simple to master
- Not just ~~FAST~~, but  _SUPER_ FAST to use

Here's how you can get started:

1. Download Alice from [this link](https://nus-cs2103-ay2324s1.github.io/website/schedule/week4/project.html).
2. Give it a gentle double-click to wake her up.
3. Start adding your tasks, she's eager to assist.
4. Sit back and let Alice manage your tasks with her adorable efficiency 😉

The best part? Alice is absolutely **FREE**!

Exciting features to look forward to:

- [x]  Task management made easy and delightful
- [ ]  Efficient deadline management (coming soon)
- [ ]  Gentle reminders to keep you on track (coming soon)

Are you a Java programmer? Alice offers a wonderful opportunity for you to brush up on your Java skills. Here's a sneak peek at the main method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Not only is Alice equipped with task management skills, but she also comes with a charming personality that is bound to make your experience pleasant and enjoyable. Give Alice a try, and let the cuteness take over your productivity journey!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

ncduy0303 committed:
Inital commit with basic greetings

ncduy0303 committed:
Add echo user inputs

ncduy0303 committed:
Add indentation in output

ncduy0303 committed:
Remove logo

ncduy0303 committed:
Add ability to store text input & display them when requested

ncduy0303 committed:
Add a new class representing tasks

ncduy0303 committed:
Add the ability to mark/unmark tasks as done

ncduy0303 committed:
Add subclasses of task

ncduy0303 committed:
Add support for tracking three types of tasks

ncduy0303 committed:
Add text testcase

ncduy0303 committed:
Add an exception class for the chatbot

ncduy0303 committed:
Handle errors for chatbot

ncduy0303 committed:
Add support for deleting tasks from the list

ncduy0303 committed:
Refactor using enum

ncduy0303 committed:
Update .gitignore

ncduy0303 committed:
Add new string format to tasks

ncduy0303 committed:
Add chatbot data storage feature

ncduy0303 committed:
Merge branch 'ncduy0303/branch-Level-7'

ncduy0303 committed:
Store real datetime objects in tasks

ncduy0303 committed:
Update test for new datetime format

ncduy0303 committed:
Merge branch 'ncduy0303/branch-Level-8'

ncduy0303 committed:
Refactor storage

ncduy0303 committed:
Refactor TaskList

ncduy0303 committed:
Add empty description exception

ncduy0303 committed:
Refactor UI

ncduy0303 committed:
Refactor Command

ncduy0303 committed:
Refactor Parser

ncduy0303 committed:
Merge branch 'ncduy0303/branch-A-MoreOOP'

ncduy0303 committed:
Divide classes into packages

ncduy0303 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

ncduy0303 committed:
Set up gradle

ncduy0303 committed:
Add some unit tests

ncduy0303 committed:
Add search task by keyword

ncduy0303 committed:
Add JavaDoc comments

ncduy0303 committed:
Reformat code using IDE

ncduy0303 committed:
Fix switch indentation

ncduy0303 committed:
Add some line breaks

ncduy0303 committed:
Merge branch 'branch-A-JavaDoc'

ncduy0303 committed:
Merge branch 'branch-A-CodingStandard'

ncduy0303 committed:
Merge branch 'branch-Level-9'

ketweeen commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -96,6 +97,8 @@ public static Command parse(String fullCommand) throws DukeException {
                 }
+            case FIND:
+                return new FindCommand(argument);
             default:
```
Nice use of default!

ketweeen commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,54 @@
+    public void run() {
+        ui.showWelcome();
+        boolean isExit = false;
+        while (!isExit) {
```
Perhaps can just do while(true) { ... }

ketweeen commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,54 @@
+        ui.showWelcome();
+        boolean isExit = false;
+        while (!isExit) {
+            try {
```
Nice try-catch statements!

ketweeen commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,54 @@
+                ui.showError(e.getMessage());
+            } finally {
+                ui.showLine();
+            }
```
Nice try-catch statements

ketweeen commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,54 @@
+        }
+    }
+
+    public static void main(String[] args) {
```
Can consider adding javadocs to constructors and main

ketweeen commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,25 @@
+        this.task = task;
+    }
+
+    @Override
```
For overriden methods, can do `{@inheritDoc}` for the Javadoc

ketweeen commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,109 @@
+ * Represents a parser that parses user input into specific commands.
+ */
+public class Parser {
+    private static enum COMMAND {
```
Write enum name in PascalCase (i.e. Command)

ncduy0303 committed:
Set up checkstyle plugin

ncduy0303 committed:
Fix styling issues

ncduy0303 committed:
Update .gitignore

ncduy0303 committed:
Merge branch 'ncduy0303/branch-A-CheckStyle'

ncduy0303 committed:
Refactor ui functions to return string output

ncduy0303 committed:
Set up GUI

ncduy0303 committed:
Add some styling

ncduy0303 committed:
Merge branch 'ncduy0303/branch-Level-10'

zsh-eng commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,79 @@
+        while (!isExit) {
+            try {
+                String fullCommand = ui.readCommand();
+                ui.showLine(); // show the divider line ("_______")
```
Perhaps consider removing the comment? One option is to rename the method to be more descriptive, for instance `showDividerLine()`.

zsh-eng commented on src/main/java/duke/command/MarkAsDoneCommand.java:
```diff
@@ -0,0 +1,24 @@
+ * Represents a command issued by the user to mark a task as done.
+ */
+public class MarkAsDoneCommand extends Command {
+    private final int index; // The index of the task to be marked as done.
```
Perhaps `taskIndex` would be more descriptive? For this and other `index` variables in the Command classes.

zsh-eng commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,109 @@
+            case BYE:
+                return new ExitCommand();
+            case TODO: {
+                String description = argument;
+                Task task = new Todo(description);
+                return new AddCommand(task);
+            }
+            case DEADLINE:
+                try {
+                    inputs = argument.split(" /", 2);
+                    String description = inputs[0];
+                    String by = inputs[1].split(" ", 2)[1];
+                    Task task = new Deadline(description, by);
+                    return new AddCommand(task);
+                } catch (ArrayIndexOutOfBoundsException e) {
+                    throw new DukeException("\u2639 OOPS!!! The deadline of a deadline cannot be empty.");
+                } catch (DateTimeParseException e) {
+                    throw new DukeException(
+                        "\u2639 OOPS!!! The date time format must be something like 2007-12-03T10:15:30.");
+                }
+            case EVENT:
+                try {
+                    inputs = argument.split(" /", 3);
+                    String description = inputs[0];
+                    String from = inputs[1].split(" ", 2)[1];
+                    String to = inputs[2].split(" ", 2)[1];
+                    Task task = new Event(description, from, to);
+                    return new AddCommand(task);
+                } catch (ArrayIndexOutOfBoundsException e) {
+                    throw new DukeException(
+                        "\u2639 OOPS!!! The start time or end time of an event cannot be empty.");
+                } catch (DateTimeParseException e) {
+                    throw new DukeException(
+                        "\u2639 OOPS!!! The date time format must be something like 2007-12-03T10:15:30.");
+                }
+            case FIND:
+                return new FindCommand(argument);
+            default:
+                return new Command();
+            }
```
Consider splitting this `switch` statement into multiple helper methods for easier readability.

For reference, see [Avoid deep nesting](https://nus-cs2103-ay2324s1.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-deep-nesting)

zsh-eng commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,130 @@
+            String description = inputs[2];
+
+            switch (taskType) {
+            case "T":
+                this.add(new Todo(description, isDone));
+                break;
+            case "D":
+                String by = inputs[3];
+                this.add(new Deadline(description, by, isDone));
+                break;
+            case "E":
+                String from = inputs[3];
+                String to = inputs[4];
+                this.add(new Event(description, from, to, isDone));
+                break;
```
Perhaps using enums here would be more descriptive?

zsh-eng reviewed (commented):
Overall, code is clean and easy to understand. Just some minor suggestions!

ncduy0303 committed:
Add more javadoc

ncduy0303 committed:
Rename Duke to Alice

ncduy0303 committed:
SLAP the parsing main logic

ncduy0303 committed:
Refactor variable names

ncduy0303 committed:
Enable assertions to `TaskList` to catch potential bugs

Currently, the `TaskList` class methods lack assertions, which means potential bugs could go undetected during runtime, leading to unexpected behaviors and potential data corruption.

The lack of assertions makes the system vulnerable to subtle bugs that can be difficult to detect and diagnose.

Using assertions to catch potential bugs is a proactive approach to software development.

ncduy0303 committed:
Improve code quality

Currently, some parts of the code base do not follow the recommended code quality guide

Changes include:
* Remove magic number
* Reduce code duplication in error messages

This will improve future maintenance of the project.

ncduy0303 committed:
Merge pull request #1 from ncduy0303/ncduy0303/branch-A-Assertions

Enable assertions to `TaskList` to catch potential bugs

ncduy0303 committed:
Merge branch 'master' into ncduy0303/branch-A-CodeQuality

ncduy0303 committed:
Merge pull request #2 from ncduy0303/ncduy0303/branch-A-CodeQuality

Improve code quality

ncduy0303 committed:
Add tagging system to tasks

ncduy0303 committed:
Merge branch 'ncduy0303/branch-C-Tagging'

ncduy0303 committed:
Add CI

ncduy0303 committed:
Merge branch 'ncduy0303/branch-A-CI'

ncduy0303 committed:
Add margin, padding, background to DialogBox

ncduy0303 committed:
Update pictures

ncduy0303 committed:
Update GUI

ncduy0303 committed:
Merge branch 'ncduy0303/branch-A-BetterGui'

ncduy0303 committed:
Improve Alice personality

ncduy0303 committed:
Merge branch 'ncduy0303/branch-A-Personality'

ncduy0303 committed:
Update tests

ncduy0303 committed:
Remove unused code

ncduy0303 committed:
Add screenshot

ncduy0303 committed:
Add exit functionality

ncduy0303 committed:
Improve FindCommand to accept multiple keywords

ncduy0303 committed:
Change BYE to EXIT command

ncduy0303 committed:
Add CLI usage option

ncduy0303 committed:
Update documentation

ncduy0303 committed:
Update README.md

ncduy0303 committed:
Update README.md

ncduy0303 committed:
Update README.md

ncduy0303 committed:
Update javafx version

ncduy0303 committed:
Update README.md
